adr/0003-multiple-templates-per-component.md

```diff
@@ -0,0 +1,52 @@
+# 1. Allow multiple templates 
+
+Date: 2021-08-30
+
+## Status
+
+Proposed.
+
+## Context
+
+As components become larger (for example, because you are implementing a whole page), it becomes
+useful to be able to extract sections of the view to a different file. ActionView has
+partials, and ViewComponent lacks a similar mechanism. 
+
+ActionView partials have the problem that their interface is not introspectable. Data
+may be passed into the partial via ivars or locals, and it is impossible to know
+which without actually opening up the file. Additionally, partials are globally 
+invocable, thus making it difficult to detect if a given partial is in use or not,
+and who are its users.
+
+An option would be to extract another Component for the extracted section, but that
+has the following drawbacks:
+
+1. It creates a new public class.
+2. If new component is invoked repeatedly (eg, for a list of items), this creates
+   GC pressure by creating lots of intermediate objects.
+
+## Decision
+
+We will allow having multiple templates in the sidecar asset. Each asset will be compiled to
+it's own method `call_<template_name>`. In order to allow the compiled method to receive arguments,
+the component must define them via a `template_arguments :template_name, :argument1, :argument2`.
```

Comment:
👋🏻 thank you for taking the time to update this PR with docs and the ADR ❤️ 

This bit is perhaps the part I'm the least comfortable with. I'm wary of re-implementing Ruby method parameters via a DSL.

I've been trying to rack my brain on how we could pull this off without a DSL, and I think this might work: we could compile additional templates to the end of methods defined in a specific format on a component. Let me try to explain (😅 it's the end of the day for me)

```rb
class MyComponent < ViewComponent::Base
  def render_other(message:)
  end
end
```

```erb
<%# my_component/other.html.erb %>

<%= message %>
```

In ViewComponent, we would then compile the template and append it to the bottom of the `render_other` method body, giving us something like:

```rb
def render_other(message:)
  @output_buffer.safe_append message
end
```

Basically, if you defined a template `other`, we'd look for a `render_other` method and compile it into that method. This approach is for sure inspired by @camertron's [rux](https://github.com/camertron/rux) ❤️ 

What do you think?

Reply:
Hmm, interesting idea. Looks doable, and has the benefit of eliminating the only-kwargs restriction. However, it feels a bit more magical, and has a few corner cases that I'm not sure how to handle:

### Should VC just overwrite the method? 

It might be confusing if I have: 
```ruby
def render_other(message)
  message.squish!
end
```

What should the resulting method be? (Is it even possible to append to a method?) Should we error out to avoid confusion? If the body is ignored, did we actually gain more by reusing the ruby method definition?

### What happens if `render_other` is defined but there is no template?

Should we error out? or just let it be?

### What happens if there is no `render_other` method?

Should we just generate an argument-less method?